### PR TITLE
No need to setify fully

### PR DESCRIPTION
**Pull request by forki**

### CODE REVIEW

**Comment by forki** [comment]
done

File: src/fsharp/lib.fs [hunk 1]
```diff
@@ -230,6 +232,14 @@ module ListSet =
     // Note: if duplicates appear, keep the ones toward the _front_ of the list
     let setify f l = List.foldBack (insert f) (List.rev l) [] |> List.rev
 
+    let hasDuplicates f l = 
+        let rec hasDuplicates acc rest =
```
Comment (dsyme) [inline src/fsharp/lib.fs]: Don't use this pattern of having an inner function with the same name as the outer function please, it's not a good coding pattern

Comment (dsyme) [reply src/fsharp/lib.fs]: Also this function looks just as quadratic as the previous function. I'd rather not make this change unless complexity changes.

Comment (smoothdeveloper) [reply src/fsharp/lib.fs]: @dsyme I might miss something but I don't think the reason for the change was the complexity when the list has no duplicates (same complexity), but to exit early if one is found instead of going all the way and compare the length.

I'd say it is now a "quadratic linear scan over a growing list" but it doesn't incur the quadratic and last calls to List.rev.

Also, at usage spot, it makes the high level intent clearer than calling setify and then comparing the length.

For the naming, I can't say, I see haskell code who would name it `go` instead.

> I'd rather not make this change unless complexity changes.

If we don't want the change in implementation, I think keeping the function (but implementing the same way it was done) is still a good change.

File: src/fsharp/lib.fs [hunk 2]
```diff
@@ -178,6 +178,8 @@ module ListAssoc =
 //------------------------------------------------------------------------
 
 module ListSet = 
+    open System.Threading
```
Comment (cartermp) [inline src/fsharp/lib.fs]: Why is this added?

Comment (forki) [reply src/fsharp/lib.fs]: removed it.

File: src/fsharp/lib.fs (hunk 1 above)
Comment (forki) [reply src/fsharp/lib.fs]: changed the function to use a hashset.

Comment (forki) [reply src/fsharp/lib.fs]: changed the function name and did some optmization for the simple cases
